orchestrator/ethereum_gravity/src/utils.rs

```diff
@@ -1,256 +1,218 @@
-use clarity::abi::{encode_call, Token};
-use clarity::Uint256;
-use clarity::{abi::encode_tokens, Address as EthAddress};
+use crate::types::EthClient;
+use ethers::core::abi::{self, Token};
+use ethers::middleware::gas_oracle::{Etherscan, GasCategory};
+use ethers::prelude::gas_oracle::GasOracle;
+use ethers::prelude::*;
+use ethers::types::Address as EthAddress;
+use ethers::utils::keccak256;
+use gravity_abi::gravity::*;
 use gravity_utils::error::GravityError;
+use gravity_utils::ethereum::{downcast_to_u64, vec_u8_to_fixed_32};
 use gravity_utils::types::*;
-use sha3::{Digest, Keccak256};
-use std::u128::MAX as U128MAX;
-use std::u64::MAX as U64MAX;
-use web30::{client::Web3, jsonrpc::error::Web3Error};
-
-// pub fn get_correct_sig_for_address(
-//     address: CosmosAddress,
-//     confirms: &[ValsetConfirmResponse],
-// ) -> (Uint256, Uint256, Uint256) {
-//     for sig in confirms {
-//         if sig.eth_signer == address {
-//             return (
-//                 sig.eth_signature.v.clone(),
-//                 sig.eth_signature.r.clone(),
-//                 sig.eth_signature.s.clone(),
-//             );
-//         }
-//     }
-//     panic!("Could not find that address!");
-// }
+use std::{cmp::min, result::Result};
 
 pub fn get_checkpoint_abi_encode(
     valset: &Valset,
     gravity_id: &str,
 ) -> Result<Vec<u8>, GravityError> {
     let (eth_addresses, powers) = valset.filter_empty_addresses();
-    Ok(encode_tokens(&[
-        Token::FixedString(gravity_id.to_string()),
-        Token::FixedString("checkpoint".to_string()),
-        valset.nonce.into(),
-        eth_addresses.into(),
-        powers.into(),
+    let eth_addresses = eth_addresses
+        .iter()
+        .map(|address| Token::Address(*address))
+        .collect();
+    let powers = powers
+        .iter()
+        .map(|power| Token::Uint((*power).into()))
+        .collect();
+
+    Ok(abi::encode(&[
+        Token::FixedBytes(gravity_id.as_bytes().to_vec()),
+        Token::FixedBytes("checkpoint".as_bytes().to_vec()),
+        Token::Uint(valset.nonce.into()),
+        Token::Array(eth_addresses),
+        Token::Array(powers),
```

Comment:
Is there a function in the generated ABI that does this? it's not fatal to call the encoder directly. More of an idle question

Reply:
I don't believe so. As far as I can tell these structures are manually created by us for use in confirmation messages and not by any types defined in the Gravity contract.

Reply:
Here is how the contract manually composes the valset update checkpoint:
https://github.com/PeggyJV/gravity-bridge/blob/3c8f2011342a61c9057ea8fe6d82c6d6aa40990d/solidity/contracts/Gravity.sol#L155-L160

Similarly for transaction batches and for logic calls, the contract is creating a hash of a manually packed ABI structure:
https://github.com/PeggyJV/gravity-bridge/blob/3c8f2011342a61c9057ea8fe6d82c6d6aa40990d/solidity/contracts/Gravity.sol#L354-L366

https://github.com/PeggyJV/gravity-bridge/blob/3c8f2011342a61c9057ea8fe6d82c6d6aa40990d/solidity/contracts/Gravity.sol#L460-L475